src/seqn/trans.lisp

```diff
@@ -0,0 +1,399 @@
+(in-package :geb.seqn.trans)
+
+(defun remove0-dom (morphism)
+  (remove 0 (dom morphism)))
+
+;; Add range-n function
+
+(defun range-constraints-dom (domain)
+  (cond ((zerop (length domain))   nil)
+        ((= (length domain) 1)    (list (geb.vampir:range-n
+                                         (vamp:make-constant :const (car domain))
+                                         (vamp:make-wire :var
+                                                         (intern (format nil "x~a" 1))))))
+        (t (cons (geb.vampir:range-n (vamp:make-constant :const (car (last domain)))
+                                (vamp:make-wire :var
+                                                (intern (format nil "x~a" (length domain)))))
+            (range-constraints-dom (butlast
+                                    domain))))))
+
+(defmethod to-circuit ((morphism <seqn>) name)
+  "Turns a SeqN term into a Vamp-IR Gate with the given name
+Note that what is happening is that we look at the domain of the morphism
+and skip 0es, making non-zero entries into wires"
+  (let* ((wire-count (length (dom morphism)))
+         (wires (loop for i from 1 to wire-count
+                      collect (vamp:make-wire :var (intern (format nil "x~a" i)
+                                                           :keyword)))))
+    (list
+     (vamp:make-alias :name name
+                      :inputs wires
+                      :body  (append (range-constraints-dom (dom morphism))
+                                     (list
+                                      (vamp:make-tuples :wires
+                                                        (remove nil (mapcar (lambda (x) (if (= (car x) 0)
+                                                                                            nil
+                                                                                            (cadr x)))
+                                                                            (prod-list (cod morphism)
+                                                                                       (to-vampir morphism
+                                                                                                  wires
+                                                                                                  nil))))
+                                                        )))))))
```

Comment:
Bad formatting, let me fix that first, before logic.

```lisp
(defmethod to-circuit ((morphism <seqn>) name)
  "Turns a SeqN term into a Vamp-IR Gate with the given name
Note that what is happening is that we look at the domain of the morphism
and skip 0es, making non-zero entries into wires"
  (let* ((wire-count (length (dom morphism)))
         (wires (loop for i from 1 to wire-count
                      collect (vamp:make-wire :var (intern (format nil "x~a" i)
                                                           :keyword)))))
    (list
     (vamp:make-alias
      :name name
      :inputs wires
      :body  (append
              (range-constraints-dom (dom morphism))
              (list
               (vamp:make-tuples
                :wires
                (remove nil
                        (mapcar (lambda (x) (if (= (car x) 0)
                                            nil
                                            (cadr x)))
                                (prod-list (cod morphism)
                                           (to-vampir morphism wires nil)))))))))))
```

Reply:
Now time for the logic changes.

Please don't do a filter map with remove and mapcar. we can be more clever.

Here are the change that you should make
```lisp
MAIN> (mapcan (lambda (x) (if (zerop (car x)) nil (list (cadr x))))
              '((0 1) (2 3) (4 5)))
(3 5)
MAIN> (filter-map (lambda (x) (if (zerop (car x)) nil (cadr x)))
                  '((0 1) (2 3) (4 5)))
(3 5)
MAIN> 
```

I.E. use filter-map or mapcan, and use `zerop` over = 0